Enrichments revamp (#550)
pupil-labs/pupil-docs: src/invisible/explainers/enrichments/reference-image-mapper.md

:::

## Validate gaze mapping

<div class="pb-4" style="display:flex;justify-content:center;">
  <v-img
    class="rounded" 
    style="margin-bottom:32px;"
    :src="require('../../../media/invisible/rim/rim-in-cloud.png')"
    max-width=90%
  >
  </v-img>
</div>

To check if gaze has been mapped successfully, use the side-by-side view:
1. Select a recording
2. Select the Reference Image Mapper Enrichment
3. Select the Scene Reference Image View
4. If you want to visualize and evaluate the 3D model generated (white dots), just turn on the Point Cloud toggle!

Now when you play back the recording you can see where gaze is mapped to on your reference image for validation

## Occlusions

<div class="pb-4" style="display:flex;justify-content:center;">
  <v-img
    class="rounded" 
    :src="require('../../../media/invisible/rim/basketball-occlusion.png')"
    max-width=400px
  >
  </v-img>
</div>

Sometimes an object will occlude the feature/object of interest. The reference image mapper may not know this, so a false positive mapping could occur. 

If you need to remove the falsely mapped data points, there are a few workarounds.

> 1. Use Events and Sections to trim out false positive sections as in the above tip.

Reviewer: ```suggestion
1. Use Events and Sections to trim out false positive sections.
```

---
description: Powerful tool to map gaze data from the scene camera onto a reference image 
permalink: /invisible/explainers/enrichments/reference-image-mapper
---

# Reference Image Mapper

> Reference Image Mapper is a powerful tool to automatically map gaze onto to features/objects in the environment.

Reviewer 2: ```suggestion
Reference Image Mapper is a powerful tool to automatically map gaze onto features/objects in the environment.
```

## Setup

<Youtube src="ygqzQEzUIS4"/>

As described in the setup video, you will need two things in addition to your eye tracking recording(s) to produce a Reference Image Mapper enrichment:
1. A reference image
2. A scanning video of the object/feature(s) taken with Pupil Invisible’s scene camera

In this guide, we will show you how the Reference Image Mapper can be used in a few different situations. This should help you get set up using the enrichment in your own testing environment!

Below, gaze is mapped in four very different environments: to a **magazine cover**, a **basketball backboard**, a **supermarket shelf**, and even a **whole building**!

<PhotoGrid :images="[
  require(`../../../media/invisible/rim/heatmap-0.jpg`),
  require(`../../../media/invisible/rim/heatmap-1.jpg`),
  require(`../../../media/invisible/rim/heatmap-2.jpg`),
  require(`../../../media/invisible/rim/heatmap-3.jpg`),
]"/>

Let's take a look at what the Reference Image and Scanning Recording look like to produce these heatmaps.

> ### 1. Magazine reading

Reviewer 2: This overview is very functional, but the spacing between image, video and text could be nicer. I think fine for now, but we should touch this again during the Neon docs update!

Author: Agree. Perhaps adding an additional row below and tuning the margins and paddings on the captions

    class="rounded" 
    style="margin-bottom:14px;"
    max-width="100%" max-height="300px" contain 
    :src="require('../../../media/invisible/rim/supermarket-img.jpg')"
  ></v-img>
    This photo captures the assortment of packagíng in the coffee aisle of a supermarket.
  </div>
  <div class="col-mcontainer">
    <h4>Scanning video</h4>
    <Youtube src="d3Yk3nKDIOQ"/>
    Notice that the scanning recording is fairly slow to reduce motion blur for this feature rich shelf.
  </div>
</div>

### 4. An entire building

<div class="mcontainer" display="grid | inline-grid">
  <div class="col-mcontainer">
    <h4>Reference image</h4>
    <v-img
    class="rounded" 
    style="margin-bottom:14px;"
    max-width="100%" max-height="300px" contain 
    :src="require('../../../media/invisible/rim/building-img.jpg')"
    ></v-img>
    This is a photo of the <i>entire</i> building
  </div>
  <div class="col-mcontainer">
    <h4>Scanning video</h4>
    <Youtube src="0U4H-uOIHlw"/>
    We take a longer scanning recording, about 2 min. The angles and distances cover what a person might see whilst walking past or standing in front of the building.
  </div>
</div>

## Scanning best practices

> 

Reviewer 2: ```suggestion
Try to follow these best practices when recording a scanning video:
```
Bullet point lists starting directly after a title look a bit ugly IMO.

- Make the recording while holding the Pupil Invisible glasses in your hand rather than wearing it on your head.

> - Record the object of interest from all possible angles and from all distances a subject may look. More diversity is better. Collecting sufficiently diverse viewpoints or a large object, like a building, may take longer than capturing a small object like a magazine.

Reviewer 2: ```suggestion
- Record the object of interest from all possible angles and from all distances a subject may view them from. More diversity is better. Collecting sufficiently diverse viewpoints for a large object, like a building, may require you to move accordingly large distances.
```

- Move the glasses slowly while recording to avoid motion blur.
- Make sure to have good contrast and that your scene lighting during scanning is similar to that during mapping.

What works and what doesn’t?

> - These examples are **feature rich**. Reference Image Mapper needs enough salient visual content to produce a successful mapping.

Reviewer 2: ```suggestion
- The scene must be **feature rich** like in the examples above. There has to be enough salient visual content to produce a successful mapping.
```

> - The scenes have relatively **static** features in the environment. If there is a lot of movement or the objects change in appearance, the mapping *can* fail.

Reviewer 2: ```suggestion
- The scene needs to have relatively **static** features in the environment. If there is a lot of movement or the objects change in appearance, the mapping *can* fail.
```
Maybe this point needs some examples to illustrate? Not sure if people would understand what "objects changing in their appearance" would mean in practice 🤔

::: tip
**Ready to go?**
Why not try replicating the above examples? Or even try it with your own use-cases! If you haven’t already, we recommend you check out the [Cloud Getting Started Guide](/invisible/getting-started/analyse-recordings-in-pupil-cloud/#analyse-recordings-in-pupil-cloud), which covers the basics of working with enrichments.
:::

> ## Validate gaze mapping

Reviewer 2: ```suggestion
## Validate the gaze mapping
```

<div class="pb-4" style="display:flex;justify-content:center;">
  <v-img
    class="rounded" 
    style="margin-bottom:32px;"
    :src="require('../../../media/invisible/rim/rim-in-cloud.png')"
    max-width=90%
  >
  </v-img>
</div>

To check if gaze has been mapped successfully, use the side-by-side view:

> 1. Select a recording
> 2. Select the Reference Image Mapper Enrichment
> 3. Select the Scene / Reference Image View
> 4. If you want to visualize and evaluate the 3D model generated (white dots), just turn on the Point Cloud toggle!

Reviewer 2: ```suggestion
1. Select a recording.
2. Select the Reference Image Mapper Enrichment.
3. Select the Scene / Reference Image View.
4. If you want to visualize and evaluate the 3D model generated (white dots), just turn on the Point Cloud toggle!
```

> Now when you play back the recording you can see where gaze is mapped to on your reference image for validation

Reviewer 2: ```suggestion
Now when you play back the recording you can see where gaze is mapped to on your reference image for validation.
```

## Occlusions

<div class="pb-4" style="display:flex;justify-content:center;">
  <v-img
    class="rounded" 
    :src="require('../../../media/invisible/rim/basketball-occlusion.png')"
    max-width=400px
  >
  </v-img>
</div>

> Sometimes an object will occlude the feature/object of interest. The reference image mapper may not know this, so a false positive mapping could occur. 

Reviewer 2: ```suggestion
Sometimes an object will occlude the feature/object of interest. The reference image mapper may not recognize this, so a false positive mapping could occur. 
```

If you need to remove the falsely mapped data points, there are a few workarounds.

1. Use Events and Sections to trim out false positive sections.
2. Manually remove the affected data points in the reference image mapper export, by finding the timestamp and deleting the row in the .csv export.

## Repetitions

> 
> **Use Sections to map portions of your recordings**
> 

Reviewer 2: ```suggestion
```
Given that this comes right after an H2 it seems repetitive to have another header.
